File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n=nums.size();
        unordered_map<int,int> mp;
        int prefixSum=0,count=0;
        mp[0]=1;
        for(int i=0;i<n;i++)
        {
            prefixSum+=nums[i];
            if(mp.find(prefixSum-k)!=mp.end()) count+=mp[prefixSum-k];
            mp[prefixSum]++;
        }
        return count;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] arr, int k) {
        int n = arr.length;
        int prefix = 0;
        
        Map<Integer,Integer> mp = new HashMap<>();
        mp.put(0,1);
        int count = 0;
        
        for(int i=0;i<n;i++) {
            prefix += arr[i];
            if(mp.containsKey(prefix - k)) {
                int j = mp.get(prefix - k);
                count+= mp.get(prefix - k);
            }
            mp.put(prefix,mp.getOrDefault(prefix,0)+1);
        }
        return count;
    }
}